A75-xds-aggregate-cluster-behavior-fixes.md

```diff
@@ -0,0 +1,312 @@
+in Envoy, priorities are normally handled by the LB policy, so this would
+have made sense as a way to use the normal logic for choosing a priority.
+
+However, it turns out that aggregate clusters don't
+actually define the LB policy in the aggregate cluster;
+instead, the aggregate cluster uses a special [cluster-provided LB
+policy](https://github.com/envoyproxy/envoy/blob/e556c237fc134fde8593c0f1ce13411455469c1b/api/envoy/config/cluster/v3/cluster.proto#L113)
+that first chooses the underlying cluster and then delegates
+to the LB policy of the underlying cluster.  This is covered in the [Envoy
+documentation](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/aggregate_cluster#aggregate-cluster),
+and you can see the Envoy implementation
+of this aggregate cluster LB policy
+[here](https://github.com/envoyproxy/envoy/blob/e556c237fc134fde8593c0f1ce13411455469c1b/source/extensions/clusters/aggregate/cluster.cc#L152).
+
+This incorrectly-perceived requirement led us to a design in which each
+of a cluster's priorities could have its configuration come from a
+different underlying cluster.  This in turn meant that features like
+outlier detection ([gRFC A50][A50]) and stateful session affinity
+(gRFCs [A55] and [A60]) were designed to work only within a specific
+priority rather than working across all priorities within a cluster.
+This limitation is causing problems for stateful session affinity and
+is likely a source of friction for outlier detection as well.
+
+This design describes how we are going to change our aggregate cluster
+implementation to solve these problems.  It also describes some changes
+to the stateful session affinity design to support it working across
+priorities.
+
+### Related Proposals:
+* [A37: xDS Aggregate and Logical DNS Clusters][A37]
+* [A50: gRPC xDS Outlier Detection Support][A50]
+* [A55: xDS-Based Stateful Session Affinity for Proxyless gRPC][A55]
+* [A60: xDS-Based Stateful Session Affinity for Weighted Clusters][A60]
+* [A56: Priority LB policy][A56]
+* [A61: IPv4 and IPv6 Dualstack Backend Support][A61]
+* [A74: xDS Config Tears (pending)][A74]
+
+[A37]: A37-xds-aggregate-and-logical-dns-clusters.md
+[A50]: A50-xds-outlier-detection.md
+[A55]: A55-xds-stateful-session-affinity.md
+[A56]: A56-priority-lb-policy.md
+[A60]: A60-xds-stateful-session-affinity-weighted-clusters.md
+[A61]: A61-IPv4-IPv6-dualstack-backends.md
+[A74]: https://github.com/grpc/proposal/pull/404
+
+## Proposal
+
+There are two main parts to this proposal: aggregate cluster changes and
+stateful session affinity changes.
+
+### Aggregate Cluster Changes
+
+Instead of representing an aggregate cluster as a single cluster whose
+priorities come from different underlying clusters, we will instead
+represent an aggregate cluster as an instance of a priority LB policy
+([gRFC A56][A56]) where each child is a cds LB policy for the underlying
+cluster.
```

Comment:
If an aggregate cluster has a child that is also an aggregate cluster, should that be represented as a deep tree, with each child of the root priority LB policy corresponding to a direct child of the root aggregate cluster, or a flat tree, with each child of the root priority LB policy corresponding to an underlying cluster?

Reply:
The intent is that the code in the `XdsDependencyManager` will recursively resolve the aggregate cluster dependency tree, using the same algorithm that we'd already implemented in the cds LB policy prior to this design.  In the complete `XdsConfig` that is passed down to the LB policies, the data for an aggregate cluster will contain the ordered, flattened list of leaf policies, ignoring any intermediate aggregate clusters.

In other words, let's say you have the following clusters:

A: aggregate cluster pointing to [B, C]
B: EDS cluster
C: aggregate cluster pointing to [D, E]
D: EDS cluster
E: EDS cluster

The `XdsConfig` data for A will indicate the leaf clusters as [B, D, E].  It will not include cluster C.

There's a note about this in the "Rationale" section.

Reply:
Got it. I didn't expect that information to only be in the Rationale section.

So, does this mean that all the root cds LB policy does is read the list of underlying cds LB configs and generate a priority LB policy with one of those configs in each priority?

Reply:
Yes, the cds policy for the aggregate cluster will generate a config for a priority policy where each child is a cds policy for the underlying leaf cluster.  However, note that the LB policy configs for the cds policies for the underlying clusters do not need to include anything other than the name of the underlying cluster, because the rest of the info will come from the relevant entry in the `XdsConfig` that is passed down via an attribute.

The fact that the cds policy receives just the list of leaf clusters is actually covered in A74, not in this gRFC.  All we're doing here is changing the way we're handling that list in the cds policy.